service/history/eventsCache.go

```diff
@@ -0,0 +1,182 @@
+		return event, nil
+	}
+
+	e.metricsClient.IncCounter(metrics.EventsCacheGetEventScope, metrics.CacheMissCounter)
+	event, err := e.getHistoryEventFromStore(domainID, workflowID, runID, firstEventID, eventID, eventStoreVersion, branchToken)
+	if err != nil {
+		e.metricsClient.IncCounter(metrics.EventsCacheGetEventScope, metrics.CacheFailures)
+		return nil, err
+	}
+
+	e.Put(key, event)
+	return event, nil
+}
+
+func (e *eventsCacheImpl) putEvent(domainID, workflowID, runID string, eventID int64, event *shared.HistoryEvent) {
+	e.metricsClient.IncCounter(metrics.EventsCachePutEventScope, metrics.CacheRequests)
+	sw := e.metricsClient.StartTimer(metrics.EventsCachePutEventScope, metrics.CacheLatency)
+	defer sw.Stop()
+
+	key := newEventKey(domainID, workflowID, runID, eventID)
+	e.Put(key, event)
+}
+
+func (e *eventsCacheImpl) deleteEvent(domainID, workflowID, runID string, eventID int64) {
+	e.metricsClient.IncCounter(metrics.EventsCacheDeleteEventScope, metrics.CacheRequests)
+	sw := e.metricsClient.StartTimer(metrics.EventsCacheDeleteEventScope, metrics.CacheLatency)
+	defer sw.Stop()
+
+	key := newEventKey(domainID, workflowID, runID, eventID)
+	e.Delete(key)
+}
+
+func (e *eventsCacheImpl) getHistoryEventFromStore(domainID, workflowID, runID string, firstEventID, eventID int64,
+	eventStoreVersion int32, branchToken []byte) (*shared.HistoryEvent, error) {
+	e.metricsClient.IncCounter(metrics.EventsCacheGetFromStoreScope, metrics.CacheRequests)
+	sw := e.metricsClient.StartTimer(metrics.EventsCacheGetFromStoreScope, metrics.CacheLatency)
+	defer sw.Stop()
+
+	var historyEvents []*shared.HistoryEvent
+	if eventStoreVersion == persistence.EventStoreVersionV2 {
+		response, err := e.eventsV2Mgr.ReadHistoryBranch(&persistence.ReadHistoryBranchRequest{
+			BranchToken:   branchToken,
+			MinEventID:    firstEventID,
+			MaxEventID:    eventID + 1,
+			PageSize:      1,
+			NextPageToken: nil,
+		})
+
+		if err != nil {
+			e.metricsClient.IncCounter(metrics.EventsCacheGetFromStoreScope, metrics.CacheFailures)
+			return nil, err
+		}
+
+		historyEvents = response.HistoryEvents
+	} else {
+		response, err := e.eventsMgr.GetWorkflowExecutionHistory(&persistence.GetWorkflowExecutionHistoryRequest{
+			DomainID: domainID,
+			Execution: shared.WorkflowExecution{
+				WorkflowId: common.StringPtr(workflowID),
+				RunId:      common.StringPtr(runID),
+			},
+			FirstEventID:  eventID,
+			NextEventID:   eventID + 1,
+			PageSize:      1,
+			NextPageToken: nil,
+		})
+
+		if err != nil {
+			e.metricsClient.IncCounter(metrics.EventsCacheGetFromStoreScope, metrics.CacheFailures)
+			return nil, err
+		}
+
+		if response.History != nil {
+			historyEvents = response.History.Events
+		}
+	}
+
+	// find history event from batch and return back single event to caller
+	var historyEvent *shared.HistoryEvent
+	for _, e := range historyEvents {
+		if e.GetEventId() == eventID {
+			historyEvent = e
+			break
+		}
+	}
+
+	return historyEvent, nil
```

Comment:
should it return error when it cannot find the event? I would just return in the for loop when event was found, and return error here when event cannot be found.